File: services/core-service/src/main/java/com/itsumori/beneaththepine/core/task/domain/Task.java
package com.itsumori.beneaththepine.core.task.domain;

import java.time.Instant;
import java.util.Objects;
import java.util.UUID;

public final class Task {
    private final UUID id;
    private final UUID userId;
    private String title;
    private int minutes;
    private TaskStatus status;
    private final UUID sourceBrainDumpId;
    private final Instant createdAt;
    private Instant updatedAt;

    public Task(
            UUID id,
            UUID userId,
            String title,
            int minutes,
            TaskStatus status,
            UUID sourceBrainDumpId,
            Instant createdAt,
            Instant updatedAt
    ) {
        this.id = Objects.requireNonNull(id, "id is required");
        this.userId = Objects.requireNonNull(userId, "userId is required");
        this.title = normalizeTitle(title);
        this.minutes = validateMinutes(minutes);
        this.status = Objects.requireNonNull(status, "status is required");
        this.sourceBrainDumpId = sourceBrainDumpId;
        this.createdAt = Objects.requireNonNull(createdAt, "createdAt is required");
        this.updatedAt = Objects.requireNonNull(updatedAt, "updatedAt is required");
    }

    public static Task create(
            UUID userId,
            String title,
            int minutes,
            UUID sourceBrainDumpId,
            Instant now
    ) {
        return new Task(
                UUID.randomUUID(),
                userId,
                title,
                minutes,
                TaskStatus.READY,
                sourceBrainDumpId,
                now,
                now
        );
    }

    public void rename(String title, Instant now) {
        ensureMutable();
        this.title = normalizeTitle(title);
        this.updatedAt = Objects.requireNonNull(now, "now is required");
    }

    public void changeMinutes(int minutes, Instant now) {
        ensureMutable();
        this.minutes = validateMinutes(minutes);
        this.updatedAt = Objects.requireNonNull(now, "now is required");
    }

    public void changeStatus(TaskStatus status, Instant now) {
        ensureMutable();
        if (status == TaskStatus.ARCHIVED) {
            throw new InvalidTaskException("Use the archive operation to archive a task");
        }
        this.status = Objects.requireNonNull(status, "status is required");
        this.updatedAt = Objects.requireNonNull(now, "now is required");
    }

    public void archive(Instant now) {
        ensureMutable();
        this.status = TaskStatus.ARCHIVED;
        this.updatedAt = Objects.requireNonNull(now, "now is required");
    }

    private void ensureMutable() {
        if (status == TaskStatus.ARCHIVED) {
            throw new ArchivedTaskException();
        }
    }

    private static String normalizeTitle(String title) {
        if (title == null) {
            throw new InvalidTaskException("Task title is required");
        }
        String normalized = title.trim();
        if (normalized.length() < 2 || normalized.length() > 280) {
            throw new InvalidTaskException("Task title must contain between 2 and 280 characters");
        }
        return normalized;
    }

    private static int validateMinutes(int minutes) {
        if (minutes < 1 || minutes > 10) {
            throw new InvalidTaskException("Task minutes must be between 1 and 10");
        }
        return minutes;
    }

    public UUID id() {
        return id;
    }

    public UUID userId() {
        return userId;
    }

    public String title() {
        return title;
    }

    public int minutes() {
        return minutes;
    }

    public TaskStatus status() {
        return status;
    }

    public UUID sourceBrainDumpId() {
        return sourceBrainDumpId;
    }

    public Instant createdAt() {
        return createdAt;
    }

    public Instant updatedAt() {
        return updatedAt;
    }
}

File: services/core-service/src/task/domain/task.ts
import { randomUUID } from 'node:crypto';
import { AppError } from '../../shared/error/app-error';

export type TaskStatus = 'ready' | 'done' | 'deferred' | 'archived';
export type Task = {
  id: string; userId: string; title: string; minutes: number; status: TaskStatus;
  sourceBrainDumpId: string | null; createdAt: Date; updatedAt: Date;
};
export type TaskUpdate = { title?: string | null; minutes?: number | null; status?: Exclude<TaskStatus, 'archived'> | null };
export type ConfirmedAction = { taskId: string; title: string; minutes: number; confirmedAt: Date };

function normalizeTitle(title: string): string {
  const normalized = title.trim();
  if (normalized.length < 2 || normalized.length > 280) {
    throw new AppError('INVALID_TASK', 'Task title must contain between 2 and 280 characters', 400);
  }
  return normalized;
}
function validMinutes(minutes: number): number {
  if (!Number.isInteger(minutes) || minutes < 1 || minutes > 10) {
    throw new AppError('INVALID_TASK', 'Task minutes must be between 1 and 10', 400);
  }
  return minutes;
}
function ensureMutable(task: Task): void {
  if (task.status === 'archived') throw new AppError('TASK_ARCHIVED', 'An archived task cannot be changed', 409);
}

export function createTask(userId: string, title: string, minutes: number, sourceBrainDumpId: string | null, now = new Date()): Task {
  return { id: randomUUID(), userId, title: normalizeTitle(title), minutes: validMinutes(minutes), status: 'ready', sourceBrainDumpId, createdAt: now, updatedAt: now };
}

export function updateTask(task: Task, input: TaskUpdate, now = new Date()): Task {
  ensureMutable(task);
  if (input.title == null && input.minutes == null && input.status == null) {
    throw new AppError('INVALID_TASK', 'At least one task field must be provided', 400);
  }
  if (input.status != null && !['ready', 'done', 'deferred'].includes(input.status)) {
    throw new AppError('INVALID_TASK', 'Use the archive operation to archive a task', 400);
  }
  return { ...task, title: input.title == null ? task.title : normalizeTitle(input.title),
    minutes: input.minutes == null ? task.minutes : validMinutes(input.minutes), status: input.status ?? task.status, updatedAt: now };
}

export function archiveTask(task: Task, now = new Date()): Task {
  ensureMutable(task);
  return { ...task, status: 'archived', updatedAt: now };
}
